src/main/java/com/algorand/algosdk/transaction/AtomicTransactionComposer.java

```diff
@@ -0,0 +1,443 @@
+                .firstValid(methodCall.fv)
+                .lastValid(methodCall.lv)
+                .fee(methodCall.fee)
+                .sender(methodCall.sender)
+                .flatFee(methodCall.flatFee)
+                .applicationId(methodCall.appID)
+                .args(encodedABIArgs)
+                .note(methodCall.note)
+                .lease(methodCall.lease);
+
+        if (methodCall.rekeyTo != null)
+            txBuilder.rekey(methodCall.rekeyTo);
+
+        Transaction tx = txBuilder.build();
+
+        tx.onCompletion = methodCall.onCompletion;
+
+        this.transactionList.addAll(tempTransWithSigner);
+        this.transactionList.add(new TransactionWithSigner(tx, methodCall.signer));
+        this.methodMap.put(this.transactionList.size() - 1, methodCall.method);
+    }
+
+    /**
+     * Finalize the transaction group and returned the finalized transactions.
+     * <p>
+     * The composer's status will be at least BUILT after executing this method.
+     */
+    public List<TransactionWithSigner> buildGroup() throws IOException {
+        if (this.status.compareTo(Status.BUILT) >= 0)
+            return this.transactionList;
+
+        if (this.transactionList.size() == 0)
+            throw new IllegalArgumentException("should not build transaction group with 0 transaction in composer");
+
+        List<Transaction> groupTxns = new ArrayList<>();
+        for (TransactionWithSigner t : this.transactionList) groupTxns.add(t.txn);
+        Digest groupID = TxGroup.computeGroupID(groupTxns.toArray(new Transaction[0]));
+        for (TransactionWithSigner tws : this.transactionList) tws.txn.group = groupID;
+
+        this.status = Status.BUILT;
+        return this.transactionList;
+    }
+
+    /**
+     * Obtain signatures for each transaction in this group. If signatures have already been obtained,
+     * this method will return cached versions of the signatures.
+     * <p>
+     * The composer's status will be at least SIGNED after executing this method.
+     * <p>
+     * An error will be thrown if signing any of the transactions fails.
+     *
+     * @return an array of signed transactions.
+     */
+    public List<SignedTransaction> gatherSignatures() throws IOException, NoSuchAlgorithmException {
+        if (this.status.compareTo(Status.SIGNED) >= 0)
+            return this.signedTxns;
+
+        List<TransactionWithSigner> txnAndSignerList = this.buildGroup();
+
+        HashMap<TxnSigner, List<Integer>> transSignerToIndices = new HashMap<>();
+        List<SignedTransaction> tempSignedTxns = new ArrayList<>();
+        for (int i = 0; i < txnAndSignerList.size(); i++) tempSignedTxns.add(null);
+
+        for (int i = 0; i < txnAndSignerList.size(); i++) {
+            TransactionWithSigner tws = txnAndSignerList.get(i);
+            if (transSignerToIndices.containsKey(tws.signer)) {
+                List<Integer> curr = transSignerToIndices.get(tws.signer);
```

Comment:
could this block be simplified to:

```suggestion
                transSignerToIndices.get(tws.signer).add(i);
```